nautobot/core/apps/__init__.py

```diff
@@ -0,0 +1,247 @@
+    registry["nav_menu"]["tabs"] = OrderedDict(
+        sorted(registry["nav_menu"]["tabs"].items(), key=lambda kv_pair: kv_pair[1]["weight"])
+    )
+
+
+class PermissionsMixin:
+    """Ensure permissions through init."""
+
+    def __init__(self, permissions=None):
+        """Ensure permissions."""
+        if permissions is not None and not isinstance(permissions, (list, tuple)):
+            raise TypeError("Permissions must be passed as a tuple or list.")
+        self.permissions = permissions
+
+
+class NavMenuTab(PermissionsMixin):
+    """
+    Ths class represents a navigation menu tab. This is built up from a name and a weight value. The name is
+    the display text and the weight defines its position in the navbar.
+
+    Links are specified as Django reverse URL strings.
+    Groups are each specified as a list of NavMenuGroup instances.
+    """
+
+    permissions = []
+    groups = []
+
+    def __init__(self, name, permissions=None, groups=None, weight=1000):
+        """Ensure tab properties."""
+        super().__init__(permissions)
+        self.name = name
+        self.weight = weight
+        if groups is not None:
+            if not isinstance(groups, (list, tuple)):
+                raise TypeError("Groups must be passed as a tuple or list.")
+            elif not all(isinstance(group, NavMenuGroup) for group in groups):
+                raise TypeError("All groups defined in a tab must be an instance of NavMenuGroup")
+            self.groups = groups
+
+
+class NavMenuGroup(PermissionsMixin):
+    """
+    Ths class represents a navigation menu group. This is built up from a name and a weight value. The name is
+    the display text and the weight defines its position in the navbar.
+
+    Links are specified as Django reverse URL strings.
+    Items are each specified as a list of NavMenuItem instances.
+    """
+
+    permissions = []
+    items = []
+
+    def __init__(self, name, items=None, weight=1000):
+        """Ensure group properties."""
+        self.name = name
+        self.weight = weight
+
+        if items is not None and not isinstance(items, (list, tuple)):
+            raise TypeError("Items must be passed as a tuple or list.")
+        elif not all(isinstance(item, NavMenuItem) for item in items):
+            raise TypeError("All items defined in a group must be an instance of NavMenuItem")
+        self.items = items
+
+
+class NavMenuItem(PermissionsMixin):
+    """
+    This class represents a navigation menu item. This constitutes primary link and its text, but also allows for
+    specifying additional link buttons that appear to the right of the item in the nav menu.
+
+    Links are specified as Django reverse URL strings.
+    Buttons are each specified as a list of NavMenuButton instances.
+    """
+
+    def __init__(self, link, link_text, permissions=None, buttons=None, weight=1000):
+        """Ensure item properties."""
+        super().__init__(permissions)
+        reverse(link)
```

Comment:
Is this just there to trigger an exception if the link isn't reversible? Since you're not actually doing anything with the value returned by `reverse`?

Reply:
yeah thats right

Reply:
Maybe add an appropriate comment in each case where you're doing this check then, since it's not obvious?